### PR TITLE
Fix chunked prefill and fix kvcache

### PR DESCRIPTION
**Pull request by yuyanpeng-google**
Fix conflict in kvcache.
* remove redundant define
* move newest version chunked prefill
* fix decoder_segement_id
* add unittest of chunked prefill

# Tests

unittest

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/inference/kvcache.py [hunk 1]
```diff
@@ -746,11 +391,16 @@ def chunked_prefill_kv_cache(self, key: Array, value: Array, decoder_segment_ids
       cached_prefill_value_vars[0].value = jax.lax.dynamic_update_slice(
           cached_value_value, value_shaped_for_cache, (next_pos, 0, 0, 0)
       )
-      cached_prefill_segment_id_var.value = decoder_segment_ids
+
+      if decoder_segment_ids is not None:
+        # decoder_segment_ids is complete prompt, while cached_prefill_segment_id is max_prefill_length
+        new_segment_id_var = jnp.zeros(cached_prefill_segment_id_var.value.shape, jnp.int32)
```
Comment (mailvijayasingh) [inline MaxText/inference/kvcache.py:397]: Is this max_prefill_length or 'chunk_size'

Comment (mailvijayasingh) [reply MaxText/inference/kvcache.py:397]: Is decoder_segment_ids need a fix here, it will also need a fix in prefill call in maxengine, can you take a look at that part? 

I think this will impact accuracy a little bit

Comment (yuyanpeng-google) [reply MaxText/inference/kvcache.py:397]: It is max_prefill_length here.
If we do `output_mask = jnp.logical_and(mask, causal_mask)` in `generate_attention_mask`, the mask which depend on decoder_segment_ids need to be max_prefill_length to match the shape of causal_mask.

The prefill call in maxengine take the complete prompt as the decoder_segment_ids. However, the prompt could be shorter than max_prefill_length using shorter prompt bucket.

**Review by gpolovets1 — APPROVED** [review]
LGTM based on Mitali & Vijay's approval.

**Review by mailvijayasingh — APPROVED** [review]
